File: ports/js/src/canonical.js
// Canonical UTF-8 encoding (spec PROTOCOL.md sec 3).
//
// Operates on native JS values: null, boolean, number (must be a safe
// integer -- see encode()), string, Array, and plain Object with string
// keys. No wrapper Value class. Object key order is explicitly sorted
// at encode time (by UTF-8 byte value), NOT relied upon from insertion
// order, since JS object key iteration order has historical quirks
// around integer-like keys that we do not want silently affecting the
// protocol's canonical bytes.

export function encode(v) {
  if (v === null) return "null";
  if (typeof v === "boolean") return v ? "true" : "false";
  if (typeof v === "number") {
    if (!Number.isInteger(v)) {
      throw new TypeError("floats are excluded from the restricted protocol value model (spec sec 2)");
    }
    if (!Number.isSafeInteger(v)) {
      throw new TypeError("integer outside safe range for this JS implementation");
    }
    return String(v);
  }
  if (typeof v === "string") return encodeString(v);
  if (Array.isArray(v)) {
    return "[" + v.map(encode).join(",") + "]";
  }
  if (typeof v === "object") {
    const keys = Object.keys(v).sort((a, b) => {
      const ba = Buffer.from(a, "utf-8");
      const bb = Buffer.from(b, "utf-8");
      return Buffer.compare(ba, bb);
    });
    const parts = keys.map((k) => encodeString(k) + ":" + encode(v[k]));
    return "{" + parts.join(",") + "}";
  }
  throw new TypeError(`value not in restricted protocol model: ${typeof v}`);
}

function encodeString(s) {
  let out = '"';
  for (const ch of s) {
    const code = ch.codePointAt(0);
    if (ch === '"') out += '\\"';
    else if (ch === '\\') out += '\\\\';
    else if (ch === '\n') out += '\\n';
    else if (ch === '\t') out += '\\t';
    else if (ch === '\r') out += '\\r';
    else if (code < 0x20) out += '\\u' + code.toString(16).padStart(4, '0');
    else out += ch;
  }
  out += '"';
  return out;
}

File: ports/python/confine_core/canonical.py
"""Canonical UTF-8 encoding (spec PROTOCOL.md sec 3).

Operates directly on Python's native restricted-model types: None, bool,
int, str, list, dict[str, ...]. No wrapper Value class -- Python's dict
already supports string keys and arbitrary nesting, and we explicitly
sort keys at encode time rather than relying on insertion order, so any
dict construction order produces the same canonical bytes.
"""


def encode(v) -> str:
    if v is None:
        return "null"
    if isinstance(v, bool):
        return "true" if v else "false"
    if isinstance(v, int):
        return str(v)
    if isinstance(v, str):
        return _encode_string(v)
    if isinstance(v, list):
        return "[" + ",".join(encode(x) for x in v) + "]"
    if isinstance(v, dict):
        keys = sorted(v.keys(), key=lambda k: k.encode("utf-8"))
        parts = [_encode_string(k) + ":" + encode(v[k]) for k in keys]
        return "{" + ",".join(parts) + "}"
    raise TypeError(f"value not in restricted protocol model: {type(v)}")


def _encode_string(s: str) -> str:
    out = ['"']
    for ch in s:
        if ch == '"':
            out.append('\\"')
        elif ch == '\\':
            out.append('\\\\')
        elif ch == '\n':
            out.append('\\n')
        elif ch == '\t':
            out.append('\\t')
        elif ch == '\r':
            out.append('\\r')
        elif ord(ch) < 0x20:
            out.append('\\u%04x' % ord(ch))
        else:
            out.append(ch)
    out.append('"')
    return ''.join(out)
